File: src/test/java/ru/mondayish/kyu8/java/SimpleCipherTest.java
package ru.mondayish.kyu8.java;

import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Random;
import java.util.stream.Collectors;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class SimpleCipherTest {

    private static final Map<String, String> REPLACE_SYMBOLS = new HashMap<String, String>() {{
        put("a", "@");
        put("b", "6");
        put("c", "c");
        put("d", "|)");
        put("e", "3");
        put("f", "|=");
        put("g", "g");
        put("h", "|-|");
        put("i", "1");
        put("j", "_!");
        put("k", "|<");
        put("l", "|");
        put("m", "m");
        put("n", "n");
        put("o", "0");
        put("p", "|*");
        put("q", "q");
        put("r", "r");
        put("s", "$");
        put("t", "7");
        put("u", "u");
        put("v", "v");
        put("w", "w");
        put("x", ")(");
        put("y", "-/");
        put("z", "-/_");
    }};

    @Test
    public void basicDecipherTests() {
        assertEquals("hello world!", SimpleCipher.decipher("|-|3||0 w0r||)!"));
        assertEquals("how are you?", SimpleCipher.decipher("|-|0w @r3 -/0u?"));
        assertEquals("this is simple test", SimpleCipher.decipher("7|-|i$ 1$ $1m|*|3 73$7"));
    }

    @Test
    public void withoutChangesDecipherTests() {
        assertEquals("hello world!", SimpleCipher.decipher("hello world!"));
        assertEquals("  some text... ", SimpleCipher.decipher("  some text... "));
        assertEquals("spring data jpa", SimpleCipher.decipher("spring data jpa"));
    }

    @Test
    public void withoutLettersDecipherTests() {
        assertEquals("", SimpleCipher.decipher(""));
        assertEquals("   ", SimpleCipher.decipher("   "));
        assertEquals("!#%..//(", SimpleCipher.decipher("!#%..//("));
    }

    @Test
    public void repeatableLettersDecipherTests() {
        assertEquals("hh.ru", SimpleCipher.decipher("|-||-|.ru"));
        assertEquals("hheeellloo", SimpleCipher.decipher("|-||-|333|||00"));
        assertEquals("qqqq wwww eee r tttt yyy", SimpleCipher.decipher("qqqq wwww 333 r 7777 -/-/-/"));
    }

    @Test
    public void sameSymbolsDecipherTests() {
        assertEquals("lxlx", SimpleCipher.decipher("|)(|)("));
        assertEquals("hhlhh", SimpleCipher.decipher("|-||-|||-||-|"));
        assertEquals("lklflpll", SimpleCipher.decipher("||<||=||*||"));
    }

    @Test
    public void randomLettersDecipherTests() {
        Random random = new Random();
        for (int i = 0; i < 100; i++) {
            StringBuilder expected = new StringBuilder();
            int expectedLength = 1 + random.nextInt(100);
            for (int j = 0; j < expectedLength; j++) {
                expected.append((char) ('a' + random.nextInt(26)));
            }

            String cipher = Arrays.stream(expected.toString().split(""))
                    .map(REPLACE_SYMBOLS::get).collect(Collectors.joining());
            assertEquals(expected.toString(), SimpleCipher.decipher(cipher));
        }
    }
}

File: src/test/java/ru/mondayish/kyu8/kotlin/SimpleCipherTest.kt
package ru.mondayish.kyu8.kotlin

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import java.util.*
import java.util.stream.Collectors
import kotlin.random.Random

internal class SimpleCipherTest {

    private val replaceSymbols: Map<String, String> = mapOf(
        Pair("a", "@"), Pair("b", "6"), Pair("c", "c"), Pair("d", "|)"), Pair("e", "3"), Pair("f", "|="),
        Pair("g", "g"), Pair("h", "|-|"), Pair("i", "1"), Pair("j", "_!"), Pair("k", "|<"), Pair("l", "|"),
        Pair("m", "m"), Pair("n", "n"), Pair("o", "0"), Pair("p", "|*"), Pair("q", "q"), Pair("r", "r"),
        Pair("s", "$"), Pair("t", "7"), Pair("u", "u"), Pair("v", "v"), Pair("w", "w"), Pair("x", ")("),
        Pair("y", "-/"), Pair("z", "-/_")
    )

    @Test
    fun basicDecipherTests() {
        assertEquals("hello world!", SimpleCipher.decipher("|-|3||0 w0r||)!"))
        assertEquals("how are you?", SimpleCipher.decipher("|-|0w @r3 -/0u?"))
        assertEquals("this is simple test", SimpleCipher.decipher("7|-|i$ 1$ $1m|*|3 73$7"))
    }

    @Test
    fun withoutChangesDecipherTests() {
        assertEquals("hello world!", SimpleCipher.decipher("hello world!"))
        assertEquals("  some text... ", SimpleCipher.decipher("  some text... "))
        assertEquals("spring data jpa", SimpleCipher.decipher("spring data jpa"))
    }

    @Test
    fun withoutLettersDecipherTests() {
        assertEquals("", SimpleCipher.decipher(""))
        assertEquals("   ", SimpleCipher.decipher("   "))
        assertEquals("!#%..//(", SimpleCipher.decipher("!#%..//("))
    }

    @Test
    fun repeatableLettersDecipherTests() {
        assertEquals("hh.ru", SimpleCipher.decipher("|-||-|.ru"))
        assertEquals("hheeellloo", SimpleCipher.decipher("|-||-|333|||00"))
        assertEquals("qqqq wwww eee r tttt yyy", SimpleCipher.decipher("qqqq wwww 333 r 7777 -/-/-/"))
    }

    @Test
    fun sameSymbolsDecipherTests() {
        assertEquals("lxlx", SimpleCipher.decipher("|)(|)("))
        assertEquals("hhlhh", SimpleCipher.decipher("|-||-|||-||-|"))
        assertEquals("lklflpll", SimpleCipher.decipher("||<||=||*||"))
    }

    @Test
    fun randomLettersDecipherTests() {
        for (i in 0..99) {
            val expected = StringBuilder()
            val expectedLength = 1 + Random.nextInt(100)
            for (j in 0 until expectedLength) {
                expected.append(('a'.code + Random.nextInt(26)).toChar())
            }
            val cipher = Arrays.stream(expected.toString().split("")
                .dropWhile { it.isEmpty() }.dropLastWhile { it.isEmpty() }
                .toTypedArray()).map { key: String? -> replaceSymbols[key] }.collect(Collectors.joining())
            assertEquals(expected.toString(), SimpleCipher.decipher(cipher))
        }
    }
}